text/0000-cargo-miri-integration.md

```diff
@@ -0,0 +1,146 @@
+
+```sh
+corgi <sysroot>/cargo/bin/build/main.rs -- /path/to/your/crate/main.rs
+```
+
+Now suppose that you want to do something complicated with your build process, like bootstrapping your program. We can do this by writing our own `cargo build` and specicifying it in `Cargo.toml` like this:
+
+```toml
+...
+
+[[cargo]]
+main = "/path/to/your/crate/cargo.rs"
+```
+
+We then create the `cargo.rs` file, which is a normal rust binary that does dependency resolution, builds everything, and links everything. It can run arbitrary rust code. If we exclude this table from the `Cargo.toml`, then the default `cargo-build` program from the standard rust distribution is used.
+
+Obviously, rewriting all of `cargo` would be a bit painful. Luckily, we don't have to. We can import `libcargo` and use the parts of the standard `cargo-build` that we don't want to replace or modify.
+
+Finally, we can do `cargo build` just like normal, except this time, `cargo build` will end up running something like:
+
+```sh
+corgi /path/to/your/crate/cargo.rs -- /path/to/your/crate/main.rs
+```
+
+One more thing: `corgi` runs everything in `miri`, so it's super slow and finds all of the UB in your build system. Awesome! This is safe, sandboxed, and discourages using too many dependencies while being infinitely customizable.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `corgi`
+
+The `corgi` tool is introduced. `corgi` functions as a virtual machine, and the build system runs inside it via `miri`. `miri` is extended to virtualize the execution of system calls, so that we can execute only system calls that won't cause UB.
+
+`corgi` terminates the build with an error if the build system contains any UB.
+
+## Translating `cargo` commands
+
+Under this RFC, all of the `cargo <foo>` subcommands become thin wrappers to call `corgi` on some rust code. Usually, that rust code will be the standard implementations of the `cargo` tools, found in the sysroot of the user's distribution. However, as shown above, the user can customize the actual implementation of their `cargo` tools by specifying the `[[cargo]]` table in their `Cargo.toml`.
+
+Commands are translated as follows:
+
+```sh
+cargo foo
+```
+
+becomes
+
+```sh
+corgi $PATH_TO_FOO -- /path/to/your/crate/main.rs
+```
+
+where `$PATH_TO_FOO` is either the path specified in `Cargo.toml` or `<sysroot>/cargo/bin/foo/main.rs` if unspecified.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The slowness of `miri` may increase build times mildly. However, this expense is deemed small in comparison to the vast might of safe-build-system-yness that will be unleashed on a UB-impoverished world.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+> - Why is this design the best in the space of possible designs?
+
+Seems obvious, right? `miri`.
```

Comment:
Yes, this is the best reason, and anyone who argues otherwise should obviously be writing their build scripts in assembler